src/main/java/net/sf/mzmine/modules/visualization/spectra/simplespectra/spectraidentification/spectraldatabase/SpectraIdentificationSpectralDatabaseTask.java

```diff
@@ -0,0 +1,192 @@
+import net.sf.mzmine.main.MZmineCore;
+import net.sf.mzmine.modules.visualization.spectra.simplespectra.SpectraPlot;
+import net.sf.mzmine.parameters.ParameterSet;
+import net.sf.mzmine.taskcontrol.AbstractTask;
+import net.sf.mzmine.taskcontrol.TaskStatus;
+import net.sf.mzmine.util.spectraldb.entry.SpectralDBEntry;
+import net.sf.mzmine.util.spectraldb.parser.AutoLibraryParser;
+import net.sf.mzmine.util.spectraldb.parser.LibraryEntryProcessor;
+
+/**
+ * Task to compare single spectra with spectral databases
+ * 
+ * @author Ansgar Korf (ansgar.korf@uni-muenster.de)
+ */
+class SpectraIdentificationSpectralDatabaseTask extends AbstractTask {
+
+  private Logger logger = Logger.getLogger(this.getClass().getName());
+
+  private final File dataBaseFile;
+
+  private ParameterSet parameters;
+
+  private Scan currentScan;
+  private SpectraPlot spectraPlot;
+
+  private List<SpectralMatchTask> tasks;
+
+  private SpectraIdentificationResultsWindow resultWindow;
+
+  private int totalTasks;
+
+  SpectraIdentificationSpectralDatabaseTask(ParameterSet parameters, Scan currentScan,
+      SpectraPlot spectraPlot) {
+
+    this.parameters = parameters;
+    dataBaseFile = parameters
+        .getParameter(SpectraIdentificationSpectralDatabaseParameters.dataBaseFile).getValue();
+    this.currentScan = currentScan;
+    this.spectraPlot = spectraPlot;
+
+  }
+
+  /**
+   * @see net.sf.mzmine.taskcontrol.Task#getFinishedPercentage()
+   */
+  @Override
+  public double getFinishedPercentage() {
+    if (totalTasks == 0 || tasks == null)
+      return 0;
+    return ((double) totalTasks - tasks.size()) / totalTasks;
+  }
+
+  /**
+   * @see net.sf.mzmine.taskcontrol.Task#getTaskDescription()
+   */
+  @Override
+  public String getTaskDescription() {
+    return "Spectral database identification of spectrum " + currentScan.getScanDefinition()
+        + " using database " + dataBaseFile;
+  }
+
+  /**
+   * @see java.lang.Runnable#run()
+   */
+  @Override
+  public void run() {
+    setStatus(TaskStatus.PROCESSING);
+    int count = 0;
+
+    // add result frame
+    resultWindow = new SpectraIdentificationResultsWindow(currentScan);
+    resultWindow.setVisible(true);
+
+    try {
+      tasks = parseFile(dataBaseFile);
+      totalTasks = tasks.size();
+      if (!tasks.isEmpty()) {
+        // wait for the tasks to finish
+        while (!isCanceled() && !tasks.isEmpty()) {
+          for (int i = 0; i < tasks.size(); i++) {
+            if (tasks.get(i).isFinished() || tasks.get(i).isCanceled()) {
+              count += tasks.get(i).getCount();
+              tasks.remove(i);
+              i--;
+            }
+          }
+          try {
+            Thread.sleep(100);
+          } catch (Exception e) {
+            cancel();
+          }
+        }
+      } else {
+        setStatus(TaskStatus.ERROR);
+        setErrorMessage("DB file was empty - or error while parsing " + dataBaseFile);
+        throw new MSDKRuntimeException(
```

Comment:
Why throw new exception here? After setting the status to ERROR, you should just ```return```.

Reply:
approved